---
title: "singlecell"
format: html
---

### Single-cell sequencing protocols

1. 3 types of protocols mainly differentiated by cell isolation method
   - Commercial Fluidigm C1 microfluidic chip based
   - Well plate based
   - Microfluidics device based 
     + Use device to capture single cell in an emulsion or droplet
     + Captures only 3' of transcript (R2 end, R1 + cell barcode + UMI at 5')
     + 10x Genomics Chromium (outperformed two other similar methods but 2x more expensive per cell, Zhang et al. 2019)
2. Potential biases
   - Protocols have biases for certain transcripts
     + "10X Genomics favored the capture and amplification of shorter genes and genes with higher GC content" - Scbp (Single-cell best practices)
   - Some cell states are more vulnerable to dissociation (e.g. non-neuronal cells survive better than neurons - Darmanis et al. 2015, Scbp)

### Raw data processing

1. Get FASTQ files (Sequence + quality)
   - Manual download from https://www.ebi.ac.uk/ena/browser/view/\<SR ID\>
     + Batch download using SRA Toolkit:
     
       ```{bash}
       #| eval: false
       fastq-dump --split-files --gzip SRR6334436
       ```
       
2. Check FASTQ sequence read quality
   - FastQC
   - MultiQC can be used to combine FastQC reports from each FASTQ

### Read mapping/alignment and quantification

Tools below can do both alignment and quantification but they can be used in combination e.g. 1 tool for alignment + 1 tool for quantification (will use BAM from alignment method).

1. Cell Ranger
   - Preparing SRA data for Cell Ranger
     + https://kb.10xgenomics.com/hc/en-us/articles/115003802691-How-do-I-prepare-Sequence-Read-Archive-SRA-data-from-NCBI-for-Cell-Ranger-
     + "Cell Ranger requires FASTQ file names to follow the `bcl2fastq` file naming convention." 
        - `[Sample Name]`\_S1_L00`[Lane Number]`\_`[Read Type]`\_001.fastq.gz
2. Kallisto
3. STARsolo (available in latest STAR versions)

### Analysis - General Remarks

1.  Check first individual datasets before integrating. Check consistency between technical and biological replicates.

### Transformations

1. Comparison and benchmarking study from Wolfgang Huber
   - Comparison to identify conceptual differences i.e. vulnerability of size factor confounding, dealing with mean-variance relationship, dealing with bimodal expression of genes
   - Benchmarking to identify performance differences
     + They measured performance based on how useful the methods are in understanding variety of cell types and states via k-NN graphs
     + They recommended benchmarking, measuring performance based on marker gene identification
   - Take-aways
     + Delta-based methods have better overall performance compared with residual-based and latent-based methods (more complicated methods)
     + Dimension reduction via PCA improves performance
     + For initial analysis of individual datasets, just try delta based methods particularly the log transformation on Seurat first
       - Avoid using CPM as it overestimates the overdispersion
       - Overdisperson of 0.05 performs well
       - Check confounding effect of size factors as delta-based methods are more vulnerable to do this compare with other types of approaches

     + Delta-based and residuals-based methods complement each other based on their conceptual differences so probably useful to see that results are robust between these two types
       - Simplest method with good performance (good starting point) is the **log(y/s +yo) with yo =1 followed by PCA**.
       
### CITE-seq / Feature barcoding and cell hashing

- CITE-seq 
  + Uses DNA-barcoded antibodies to measure surface proteins simultaneously with RNA
  + Before partitioning cells in GEMs, cells get tagged by DNA-barcoded antibodies targeting markers
  + DNA barcode will then be treated just like the other molecules 
  + So on top of getting RNA profile of single cells, with CITE-seq we can get information on expression of surface protein markers per cell
  + Amount of DNA barcode coming from antibody will reflect also level of surface protein
  + Library construction -  size selection is used to separate the amplified cDNA molecules for 3' Gene Expression and Cell Surface Protein Library Construction (product giude, Chromium Next GEM Single Cell 3' HT Reagent Kits v3.1 (Dual Index))
    - Separate library construction because the DNA barcode attached to antibody already have the sequencing read 2 (library specifc, used for PCR amplication?), meanwhile for the RNA first strands, the sequencing read 2 still have to be ligated to complete the fragment for sequencing

- Cell hashing
  + The same principle as CITE-seq but because the DNA-barcoded antibodies are used to tag different samples, antibodies can just target some ubiquitously expressed surface protein (across samples) instead of specific marker proteins
  
  
# Analysis in Seurat 

## Demultiplexing

### HTODemux()

- Starts with clustering cells based on HTO vectors
- `HTODemux(positive.quantile)` - Per HTO, identifies the cluster of cells with lowest average for the HTO, fits a negative binomial distribution to values of that cluster and then identifies the value at given positive.quantile (this value is outputed per HTO while HTODemux is running), if a cell has higher HTO value that this value at that quantile, cell is considered to be positive for that HTO

## Normalisation

- Motivation 
  + Account for systematic differences across samples or cells e.g. differences in sequencing depth / library size
  
### Normalisation methods

- LogNormalise
  + Per cell, counts are divided by a cell-specific size factor proportional to the library size
  + Then log transformation is done to convert expression data, often right skewed (closer to a log normal distribution), into a distribution that is closer to a normal distribution, also reducing impact of outlier values
- Centered log-ratio (CLR)
  + Per cell, values are transformed such that each proportion of features sum up to 1 (100%) 
  + Per cell, take the natural logarithm of the ratio of feature values to their geometric mean 
  + Transformed values are then centered per FEATURE i.e. unlike logNormalise and RC, this method does transformation across cells
- Relative counts (RC)
  + logNormalise but skipping log transformation
  
## Find variable features

- Point of methods is to stabilize the variance across the mean expression of genes, making it easier to identify genes whose variability is not simply due to mean expression

### Methods

- Variance stabilising transformation (vst)